### DISCUSSION TITLE
high block output when idling

### DISCUSSION THREAD

**Discussion by helmut72** [minio/minio#15705] [Q&A] [answered]
## NOTE
I have installed Minio in a Docker container, created a bucket, uploaded a file and let Minio running without any access. After idling Minio for some hours, block output in Docker (`docker stats minio`, column `block i/o`) is growing. Even my SQL servers don't write that much when in use (private server, few users).

## Expected Behavior
When Minio idles it shouldn't write that much to disk.

## Current Behavior
Minio write huge data (compared to my other 40+ containers) to disk when idling.

## Steps to Reproduce (for bugs)
Install Minio: https://docs.min.io/minio/baremetal/installation/deploy-minio-single-node-single-drive.html#deploy-containerized-single-node-single-drive-minio

1. Create a bucket
2. wait for hours, days
3. check `block i/o` in `docker stats minio`

## Context
Wrting somethin to disk isn't helpful to avoid wearing of a SSD.

## Your Environment

* all Minio versions published this year
* Server setup and configuration: Ubuntu 22.04, latest Docker. But also with Ubuntu 20.04 and Docker early this year. Maybe also some older versions, don't use Minio that long

**Comment by klauspost** [comment, marked as answer]
Minio runs a scanner that continuously checks your data.

Block I/O doesn't tell you if it is reads or writes. Reads are expected (also when idle, and it doesn't wear your SSD) and state is consisted during scans, IIRC every minute.

Executing `mc admin config get ALIAS scanner` will show your scanner config. You can make `delay` and `max_wait` bigger if you want to slow the scanner down. For both bigger values gives slower scanning, with `delay` being the main controller.

Use `mc admin scanner info ALIAS` to see scanner speed.

**Reply by helmut72** [reply]
Thanks for the info about scanner. I just rely on the Docker documentation: [](https://docs.docker.com/engine/reference/commandline/stats/)
"The amount of data the container has read to and **written** from block devices on the host"

Of course I don't care about reading from a scanner service and don't want to have any side effects if I increase the numbers, but being confused what will be written if Minio just idles with some scan activities every minute. All "higher" (just a private server) block i/o numbers of my other containers make sense.

**Reply by helmut72** [reply]
https://github.com/minio/minio/issues/13893

This also shows that constant write occurs, even when Minio idles. How can this be stopped? Again, I don't care about reading, but writing. Dockers Block I/O tells reading **and** writing.

I also set higher values, but this have side effects as you mentioned in the linked issue and doesn't help avoid constant writes to SSD even when Minio just idles.

Would be great to unmark as answer because it doesn't answer my question/issue. Thank you.

**Reply by klauspost** [reply]
@helmut72 It answers your question. The documentation you link says it is reads and writes, so assuming it is one or the other seems like speculation.

Either way, we have no plans to make changes to this. If you want to submit a PR for a change feel free, as long as it doesn't degrade performance or adds rarely used configuration that we will need to support. 

If you have a commercial case for this contact `sales@min.io` and we can go into a deeper discussion about your case.

**Reply by helmut72** [reply]
Thank you for your fast reply. `docker stats` outputs 2 numbers in block i/o column, one for read, one for write. And writing numbers grows even when Minio just idles. My question was how to avoid this. It looks like it's by design.

No commercial case, I just take care of my private SSD. When some process is used and writes, of course I also don't care. But I take care if something just idles but in my opinion unnecessary writes (Minio is just a single instance). My Samba service is used for AD/LDAP only. It runs for 2 weeks and writing number is a lot lower compared to Minio, which runs for 2 hours.

<img width="751" alt="Bildschirmfoto 2022-09-25 um 13 30 07" src="https://user-images.githubusercontent.com/1388751/192141254-c732f466-96c9-4f6c-85d7-16e7837af4fa.png">

Also I'm not automatically a developer if I ask something. I just need a S3 server because the used app requires it.

**Reply by JensVial** [reply]
Hello,
I habe the same issue. running minio as docker -> 4 instances with 2 drives
I confugure the scanner with mc admin -> delay = 30 and max_wait 15s but if I check the disk usage with htop and then the READ-COLUMN I see that there are reads ~2 - 5 B/s all 2-5 secs 
So the scanner option do not help in my case
This problem prevent my disks to go in energy-save mode
Is there a solution for this?

**Reply by klauspost** [reply]
@JensVial The scanner must run for maintenance operations to complete. You cannot expect drives to go into power saving mode. If you want that, you must shut down the cluster.

`delay=30` just means it does everything 30x slower than default. It will not stop.

**Reply by helmut72** [reply]
@klauspost 

> cluster

Why Minio needs scanning (reading) in a Single-Node Single-Drive (SNSD) setup? What is Minio writing when it's just scanning (reading) and there are no changes?

**Reply by klauspost** [reply]
Usage calculation, all ilm operations, integrity checks are performed.

**Reply by helmut72** [reply]
Thanks. What data will be written to disk in a SNSD setup when there are no changes?

There is some more context need. I guess your answer is to my first question.

- usage calculation: IMHO shouldn't writing anything to disk.
- ilm operations: IMHO nothing todo, because my instance just idles.
- integrity check: I guess this reads a checksum from disk and compares to stored files. IMHO again just reading.

Some of my used apps supports WebDAV, others unfortunately S3 only. Therefore I need a simple S3 endpoint/server. Both, a WebDAV and a S3 server, are more or less webservers when it comes to read and write files and let Minio running in a SNSD setup.

My WebDAV container (called "sync") runs for months and is writing, Minio (2nd instance for testing) was restarted 2 days ago and idles since then. These are the stats:

![minio](https://user-images.githubusercontent.com/1388751/194704312-298138e8-76c1-4871-ada1-98a2f21bf194.jpg)

When I read this numbers and compare it to other running containers, I always start asking what exactly will be written when it's just idling. Especially in a SNSD setup. I just want to know to understand this huge write operations.

**Reply by helmut72** [reply]
> Block I/O doesn't tell you if it is reads or writes.

Now I got a spare PC and use smartctl tool on the host, which reads the stats of the raw SSD. Started only MinIO with few buckets and files (35 files, 60MB in size), no S3 or admin access since yesterday = MinIO just idles (the whole PC too).

This is the result:

Mon 12:09 am

```
smartctl data written: 51335,251 GB

docker stats:
CONTAINER ID   NAME      CPU %     MEM USAGE / LIMIT     MEM %     NET I/O          BLOCK I/O        PIDS
b99385add0eb   minio     0.79%     279.3MiB / 15.31GiB   1.78%     665kB / 71.3MB   2.87MB / 338MB   10
```

Mon 06:49 pm

```
smartctl data written: 51336,632 GB

docker stats:
CONTAINER ID   NAME      CPU %     MEM USAGE / LIMIT   MEM %     NET I/O          BLOCK I/O         PIDS
b99385add0eb   minio     0.93%     281MiB / 15.31GiB   1.79%     666kB / 71.3MB   11.1MB / 1.35GB   10
```

Tue 07:55 am

```
smartctl data written: 51339,356 GB

docker stats:
CONTAINER ID   NAME      CPU %     MEM USAGE / LIMIT   MEM %     NET I/O          BLOCK I/O         PIDS
b99385add0eb   minio     0.00%     282MiB / 15.31GiB   1.80%     668kB / 71.3MB   27.2MB / 3.35GB   10
```

IMHO block I/O (especially it's the O in I/O = output = writes) correlates with the numbers from smartctl and block I/O in docker stats command is a valid value.

**Reply by mishak87** [reply]
I have same issue with running Minio SDSN on ZFS pool as storage for various backups it keeps writing to disk every minute or two despite nothing begin read or written.

The simple use-case for Minio to run S3 api for a disk folder for infrequent writes is degraded because of it.
Imagine running this on RaspberryPi where it will silently degrade the SD card.

**Reply by mishak87** [reply]
I suspect this is the culprit will try to make PR when I find where it is in code.
2049B written every 15s...
```
[pid 933526] mkdirat(AT_FDCWD, "/export/.minio.sys/tmp", 0777 <unfinished ...>
[pid 933491] futex(0xc00008a548, FUTEX_WAIT_PRIVATE, 0, NULL <unfinished ...>
[pid 933526] <... mkdirat resumed>)     = -1 EEXIST (File exists)
[pid 933526] openat(AT_FDCWD, "/export/.minio.sys/tmp/9a65388f-b5e8-4198-b3b6-1ce3bcf07f5b", O_WRONLY|O_CREAT|O_TRUNC|O_DSYNC|O_CLOEXEC, 0666) = 9
[pid 933526] fcntl(9, F_GETFL)          = 0x9001 (flags O_WRONLY|O_DSYNC|O_LARGEFILE)
[pid 933526] fcntl(9, F_SETFL, O_WRONLY|O_NONBLOCK|O_DSYNC|O_LARGEFILE) = 0
[pid 933526] epoll_ctl(4, EPOLL_CTL_ADD, 9, {events=EPOLLIN|EPOLLOUT|EPOLLRDHUP|EPOLLET, data={u32=462947784, u64=9214040237319652808}}) = -1 EPERM (Operation not permitted)
[pid 933526] fcntl(9, F_GETFL)          = 0x9801 (flags O_WRONLY|O_NONBLOCK|O_DSYNC|O_LARGEFILE)
[pid 933526] fcntl(9, F_SETFL, O_WRONLY|O_DSYNC|O_LARGEFILE) = 0
[pid 933526] write(9, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 2049) = 2049
[pid 933526] close(9)                   = 0
[pid 933526] openat(AT_FDCWD, "/export/.minio.sys/tmp/9a65388f-b5e8-4198-b3b6-1ce3bcf07f5b", O_RDONLY|O_NOATIME|O_CLOEXEC) = 9
[pid 933526] fcntl(9, F_GETFL)          = 0x48000 (flags O_RDONLY|O_LARGEFILE|O_NOATIME)
[pid 933526] fcntl(9, F_SETFL, O_RDONLY|O_NONBLOCK|O_LARGEFILE|O_NOATIME) = 0
[pid 933526] epoll_ctl(4, EPOLL_CTL_ADD, 9, {events=EPOLLIN|EPOLLOUT|EPOLLRDHUP|EPOLLET, data={u32=462947785, u64=9214040237319652809}}) = -1 EPERM (Operation not permitted)
[pid 933526] fcntl(9, F_GETFL)          = 0x48800 (flags O_RDONLY|O_NONBLOCK|O_LARGEFILE|O_NOATIME)
[pid 933526] fcntl(9, F_SETFL, O_RDONLY|O_LARGEFILE|O_NOATIME) = 0
[pid 933526] fstat(9, {st_mode=S_IFREG|0644, st_size=2049, ...}) = 0
[pid 933526] read(9, "\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0"..., 2049) = 2049
[pid 933526] close(9)
```

**Reply by mishak87** [reply]
It should be possible to disable it by setting env `_MINIO_DRIVE_ACTIVE_MONITORING=off` https://github.com/minio/minio/blob/1aa8896ad69a3e23b0f439cb83f873956ed620a7/cmd/globals.go#L429
Will test tomorrow.

**Reply by harshavardhana** [reply]
Locking conversation here, monitoring a drive is a production requirement and turning it off is your own risk.

So please do what makes it okay for you. But we don't recommend it.
